examples/virtio/blk_driver_vmm.c

```diff
@@ -0,0 +1,143 @@
+/* Data for the device tree to be passed to the kernel. */
+extern char _guest_dtb_image[];
+extern char _guest_dtb_image_end[];
+/* Data for the initial RAM disk to be passed to the kernel. */
+extern char _guest_initrd_image[];
+extern char _guest_initrd_image_end[];
+/* Microkit will set this variable to the start of the guest RAM memory region. */
+uintptr_t guest_ram_vaddr;
+
+/* sDDF block */
+#define BLOCK_CH 1
+#if defined(BOARD_odroidc4)
+#define SD_IRQ 222
+#elif defined(BOARD_qemu_arm_virt)
+#define BLOCK_IRQ 79
+#endif
+
+// @ericc: autogen
+#define UIO_IRQ 50
+#define UIO_CH 3
+
+void uio_ack(size_t vcpu_id, int irq, void *cookie)
+{
+    microkit_notify(UIO_CH);
+}
+
+void init(void)
+{
+    /* Initialise the VMM, the VCPU(s), and start the guest */
+    LOG_VMM("starting \"%s\"\n", microkit_name);
+    /* Place all the binaries in the right locations before starting the guest */
+    size_t kernel_size = _guest_kernel_image_end - _guest_kernel_image;
+    size_t dtb_size = _guest_dtb_image_end - _guest_dtb_image;
+    size_t initrd_size = _guest_initrd_image_end - _guest_initrd_image;
+    uintptr_t kernel_pc = linux_setup_images(guest_ram_vaddr,
+                                             (uintptr_t) _guest_kernel_image,
+                                             kernel_size,
+                                             (uintptr_t) _guest_dtb_image,
+                                             GUEST_DTB_VADDR,
+                                             dtb_size,
+                                             (uintptr_t) _guest_initrd_image,
+                                             GUEST_INIT_RAM_DISK_VADDR,
+                                             initrd_size
+                                            );
+    if (!kernel_pc) {
+        LOG_VMM_ERR("Failed to initialise guest images\n");
+        return;
+    }
+
+    /* Initialise the virtual GIC driver */
+    bool success = virq_controller_init(GUEST_VCPU_ID);
+    if (!success) {
+        LOG_VMM_ERR("Failed to initialise emulated interrupt controller\n");
+        return;
+    }
+
+    /* Register the UIO IRQ */
+    virq_register(GUEST_VCPU_ID, UIO_IRQ, uio_ack, NULL);
+
+#if defined(BOARD_odroidc4)
+    /* Register the SD card IRQ */
+    virq_register_passthrough(GUEST_VCPU_ID, SD_IRQ, BLOCK_CH);
+#if defined(DEBUG_BLK_DRIVER_VM)
+    virq_register_passthrough(GUEST_VCPU_ID, 225, 10);
+#endif
+#endif
+
+#if defined(BOARD_qemu_arm_virt)
+    /* Register the SD card IRQ */
+    virq_register_passthrough(GUEST_VCPU_ID, BLOCK_IRQ, BLOCK_CH);
+#if defined(DEBUG_BLK_DRIVER_VM)
+    virq_register_passthrough(GUEST_VCPU_ID, 33, 10);
+#endif
+#endif
+
+    /* Finally start the guest */
+    guest_start(GUEST_VCPU_ID, kernel_pc, GUEST_DTB_VADDR, GUEST_INIT_RAM_DISK_VADDR);
+}
+
+void notified(microkit_channel ch)
+{
+    bool handled = false;
+
+    handled = virq_handle_passthrough(ch);
+
+    if (ch == UIO_CH) {
+        virq_inject(GUEST_VCPU_ID, UIO_IRQ);
```

Comment:
Check return value

Reply:
Doesn't need to be an error but place an assert at least since it's unlikely to fail

Reply:
until the VGIC problem #56 is fixed this is going to crash the program. I'll just log an error for now.